Feature/libfcrepo 58

jgottwig opened the pull request:
LIBFCREPO-58. IP address to security principal filter for Tomcat.
- Create a Tomcat valve that adds an HTTP header to the request based on the IP address of the request

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Pushing first build.
-Pull properties from context
-Check header
-Overall application structure
-Plenty of test code
-Largely untested

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Updated pom file

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Sketching out more header functionality.
- Also added more error logging

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Removing unneeded dependency.
-Briefly considered working with CSVs
-Then learned more about properties files structures

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Updating comments.
-Removed question
-Added description

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Adding header functionality.
-Includes library addition to pom.xml

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Enhancements; fixes
-Classpath issue
-Caching properties
-Additional repairs

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. JUnit Tests and code tweaks.
- Attempted Mockito, but it doesn't work with Coyote Requests
- What gives?
- Added test properties file
- Refactored code to better work with tests

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Failed attempts at improving properties.
- Pushing to have available in VC.

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Refactoring class path.
- Whoopse.
- Including a few comment tweaks.
- Also restored one additional log entry.

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Pom adjustment for Apache Config
- This may not get used.

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Improving JUnit Tests.
- Plus tweaked some comments

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Fixing one comment.

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Removing properties check in invoke()
- Also adjusted tests appropriately.

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Removing unused dependency.

https://issues.umd.edu/browse/LIBFCREPO-58

peichman-umd commented:
If I add a single IP address instead of a subnet in the mapping file, the valve throws an exception on every request:

```
java.lang.IllegalArgumentException: Could not parse [192.168.40.1]
    org.apache.commons.net.util.SubnetUtils.calculate(SubnetUtils.java:275)
    org.apache.commons.net.util.SubnetUtils.<init>(SubnetUtils.java:51)
    edu.umd.lib.tomcat.valves.IPAddressMapper.getApprovals(IPAddressMapper.java:126)
    edu.umd.lib.tomcat.valves.IPAddressMapper.invoke(IPAddressMapper.java:208)
    edu.umd.lib.tomcat.valves.OptionalSSLAuthenticator.invoke(OptionalSSLAuthenticator.java:40)
    edu.umd.lib.tomcat.valves.OptionalBasicAuthenticator.invoke(OptionalBasicAuthenticator.java:99)
    org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:98)
    org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:950)
    org.apache.catalina.authenticator.SingleSignOn.invoke(SingleSignOn.java:339)
    org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:408)
    org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1040)
    org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:607)
    org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:313)
    java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    java.lang.Thread.run(Thread.java:745)
```

jgottwig commented:
I've written something to catch this exception, but is the preferred behavior to instead support non-CIDR notation?

jgottwig committed:
LIBFCREPO-58. Catching subnet exception.
- Wrapping around a try/catch
- Also added test IP to testing properties file

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Adding support for comparing single IPs.
- Also added additional junit test.

https://issues.umd.edu/browse/LIBFCREPO-58

jgottwig committed:
LIBFCREPO-58. Tweaking logic.
- Checking for valid single IP first
- This should limit exceptions

https://issues.umd.edu/browse/LIBFCREPO-58

peichman-umd commented:
One last code style comment. In a couple places you declare and initialize a variable to `null` outside of a block, but then only assign and use that variable within that block. In those cases, it would be better to do the declaration and final assignment inside the block. I am going to mark the cases of this I see via line-by-line comments on this PR.

But otherwise, this is terrific work!

peichman-umd commented on src/main/java/edu/umd/lib/tomcat/valves/IPAddressMapper.java:
```diff
@@ -0,0 +1,233 @@
+      /**
+       * Inject the header with value if the user's IP meets the above criteria.
+       */
+      String finalHeaders = null;
```
Move this declaration of `finalHeaders` inside the block and combine with its assignment.

peichman-umd commented on src/main/java/edu/umd/lib/tomcat/valves/IPAddressMapper.java:
```diff
@@ -0,0 +1,233 @@
+    Enumeration<?> propertyNames = properties.propertyNames();
+
+    List<String> approvals = new ArrayList<String>();
+    SubnetUtils utils; // Our comparison library
```
Move this declaration of `utils` inside the block and combine with its assignment (on line 132)

jgottwig committed:
LIBFCREPO-58.  Moved vars into code blocks.
- Added final assignment

https://issues.umd.edu/browse/LIBFCREPO-58
